src/analysis/inner_product_space/pi_L2.lean

```diff
@@ -150,54 +150,198 @@ begin
+end
+
+protected lemma repr_apply_apply (b : orthonormal_basis ι 𝕜 E) (v : E) (i : ι) :
+  b.repr v i = ⟪b i, v⟫ :=
 begin
-  intros x y,
-  let p : euclidean_space 𝕜 ι := v.equiv_fun x,
-  let q : euclidean_space 𝕜 ι := v.equiv_fun y,
-  have key : ⟪p, q⟫ = ⟪∑ i, p i • v i, ∑ i, q i • v i⟫,
-  { simp [sum_inner, inner_smul_left, hv.inner_right_fintype] },
-  convert key,
-  { rw [← v.equiv_fun.symm_apply_apply x, v.equiv_fun_symm_apply] },
-  { rw [← v.equiv_fun.symm_apply_apply y, v.equiv_fun_symm_apply] }
+  rw [← b.repr.inner_map_map (b i) v, b.repr_self _ _ _, euclidean_space.inner_single_left],
+  simp only [one_mul, eq_self_iff_true, map_one],
 end
 
-@[simp] lemma basis.coe_isometry_euclidean_of_orthonormal
+@[simp]
+protected lemma orthonormal (b : orthonormal_basis ι 𝕜 E) : orthonormal 𝕜 b :=
+begin
+  classical,
+  rw orthonormal_iff_ite,
+  intros i j,
+  rw [← b.repr.inner_map_map (b i) (b j), b.repr_self _ _ _, b.repr_self _ _ _],
+  rw euclidean_space.inner_single_left,
+  rw euclidean_space.single_apply,
+  simp only [mul_boole, eq_self_iff_true, map_one],
+end
+
+protected lemma sum_repr_symm (b : orthonormal_basis ι 𝕜 E) (v : euclidean_space 𝕜 ι) :
+  ∑ i , v i • b i = (b.repr.symm v) :=
+begin
+  classical,
+  have : b.repr (∑ i, v i • b i) = v :=
+  begin
+    have : ⇑(b.repr) = (b.repr.to_linear_isometry.to_linear_map) :=
+    begin
+      simp only [linear_isometry.coe_to_linear_map, fun_like.coe_fn_eq,
+      linear_isometry_equiv.coe_to_linear_isometry, eq_self_iff_true],
+    end,
+  rw this,
+  simp_rw [linear_map.map_sum, linear_map.map_smul, ← this],
+  ext i,
+  change (∑ (c : ι), (λ x, v x • (b.repr) (b x)) c) i = v i,
+  rw [@fintype.sum_apply _ _ ι _ _ i (λ x, v x • (b.repr) (b x))],
+  simp only [algebra.id.smul_eq_mul, orthonormal_basis.repr_self, pi.smul_apply,
+  finset.sum_congr],
+  simp_rw [euclidean_space.single, set.indicator, mul_ite, mul_zero],
+  simp only [mul_one, finset.mem_univ, if_true, set.mem_singleton_iff, eq_self_iff_true,
+  finset.sum_congr, finset.sum_ite_eq],
+  end,
+  conv_rhs
+  begin
+  rw ← this,
+  end,
+  simp only [linear_isometry_equiv.symm_apply_apply, eq_self_iff_true],
+end
+
+variables {ι} {𝕜} {E}
+variable {v : ι → E}
+
+
+/-- A basis that is orthonormal is an orthonormal basis. -/
+def _root_.basis.to_orthonormal_basis (v : basis ι 𝕜 E) (hv : orthonormal 𝕜 v) :
+  orthonormal_basis ι 𝕜 E :=
+  orthonormal_basis.of_repr $
+  v.equiv_fun.isometry_of_inner
+  begin
+    intros x y,
+    let p : euclidean_space 𝕜 ι := v.equiv_fun x,
+    let q : euclidean_space 𝕜 ι := v.equiv_fun y,
+    have key : ⟪p, q⟫ = ⟪∑ i, p i • v i, ∑ i, q i • v i⟫,
+    { simp [sum_inner, inner_smul_left, hv.inner_right_fintype] },
+    convert key,
+    { rw [← v.equiv_fun.symm_apply_apply x, v.equiv_fun_symm_apply] },
+    { rw [← v.equiv_fun.symm_apply_apply y, v.equiv_fun_symm_apply] }
+  end
+
+@[simp] lemma _root_.basis.coe_isometry_euclidean_of_orthonormal
```

Comment:
This lemma should probably change its name now to `basis.coe_to_orthonormal_basis_repr`, and the next to `basis.coe_to_orthonormal_basis_repr_symm`.